Code/GraphMol/Canon.cpp

```diff
@@ -1250,6 +1256,796 @@ void canonicalizeFragment(ROMol &mol, int atomIdx,
+
+  auto firstChiralVal1 =
+      allRankedValues[0].getChiralAtoms()[index1].getChiralType();
+  auto firstChiralVal2 =
+      allRankedValues[0].getChiralAtoms()[index2].getChiralType();
+
+  for (auto &rankedValue : allRankedValues) {
+    if ((firstChiralVal1 ==
+         rankedValue.getChiralAtoms()[index1].getChiralType()) !=
+        (firstChiralVal2 ==
+         rankedValue.getChiralAtoms()[index2].getChiralType())) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool doAtomAndBondVaryTheSame(const OrderedSet<RankedValue> &allRankedValues,
+                              unsigned int atomIndex1,
+                              unsigned int bondIndex2) {
+  PRECONDITION(allRankedValues.size() != 0, "bad allMols size");
+  PRECONDITION(allRankedValues.begin()->getNumChiralAtoms() > atomIndex1,
+               "atomIndex1 out of range");
+  PRECONDITION(allRankedValues.begin()->getNumChiralBonds() > bondIndex2,
+               "bondIndex2 out of range");
+
+  auto firstChiralVal1 =
+      allRankedValues[0].getChiralAtoms()[atomIndex1].getChiralType();
+  auto firstStereoVal2 =
+      allRankedValues[0].getChiralBonds()[bondIndex2].getStereoType();
+
+  for (auto &rankedValue : allRankedValues) {
+    if ((firstChiralVal1 ==
+         rankedValue.getChiralAtoms()[atomIndex1].getChiralType()) !=
+        (firstStereoVal2 ==
+         rankedValue.getChiralBonds()[bondIndex2].getStereoType())) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool doTwoBondsVaryTheSame(const OrderedSet<RankedValue> &allRankedValues,
+                           unsigned int bondIndex1, unsigned int bondIndex2) {
+  PRECONDITION(allRankedValues.size() != 0, "bad allMols size");
+  PRECONDITION(allRankedValues.begin()->getNumChiralBonds() > bondIndex1,
+               "atomIndex1 out of range");
+  PRECONDITION(allRankedValues.begin()->getNumChiralBonds() > bondIndex2,
+               "bondIndex2 out of range");
+
+  auto firstStereoVal1 =
+      allRankedValues[0].getChiralBonds()[bondIndex1].getStereoType();
+  auto firstStereoVal2 =
+      allRankedValues[0].getChiralBonds()[bondIndex2].getStereoType();
+
+  for (auto &rankedValue : allRankedValues) {
+    if ((firstStereoVal1 ==
+         rankedValue.getChiralBonds()[bondIndex1].getStereoType()) !=
+        (firstStereoVal2 ==
+         rankedValue.getChiralBonds()[bondIndex2].getStereoType())) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+void canonicalizeStereoGroups_internal(
+    std::unique_ptr<RWMol> &mol, RDKit::StereoGroupType stereoGroupType,
+    StereoGroupAbsOptions outputAbsoluteGroups) {
+  // this expanded a mol with stereo groups to a vector of values that have no
+  // stereo groups, then determines the stereo groups from that set
+  //
+
+  auto origStereoGroups =
+      mol->getStereoGroups();  // to retorore mol is an error is detected
+  try {
+    std::vector<unsigned int> ranks(mol->getNumAtoms());
+
+    mol->updatePropertyCache(true);
```

Comment:
Why is this here? It doesn't look like you have done anything to change the valence states of any atoms.

Reply:
Leaving this out caused a crash for a reaction test.

Reply:
I added a test to see if the update is needed, and this has been moved to the inital canonicalizeStereoGroups function so that it is only done once.